lib/streamlit/web/server/oauth_authlib_routes.py

```diff
@@ -0,0 +1,122 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+
+import tornado.web
+
+from streamlit.runtime.secrets import secrets_singleton
+from streamlit.web.server.oidc_mixin import TornadoOAuth
+
+
+class AuthCache:
+    def __init__(self):
+        self.cache = {}
+
+    def get(self, key):
+        return self.cache.get(key)
+
+    def set(self, key, value, expires_in):
+        self.cache[key] = value
+
+    def get_dict(self):
+        return self.cache
+
+    def delete(self, key):
+        self.cache.pop(key, None)
+
+
+my_cache = AuthCache()
+
+# Todo [kajarenc] initialize Auth clients lazily
+# todo [kajarenc] use create_oauth_client function that initialize and return client
+
+if secrets_singleton.load_if_toml_exists():
+    auth_section = secrets_singleton.get("auth")
+    if auth_section:
+        redirect_uri = auth_section.get("redirect_uri", None)
+        config = dict(auth_section)
+    else:
+        config = {}
+        redirect_uri = "/"
+else:
+    config = {}
+    redirect_uri = "/"
+
+oauth = TornadoOAuth(config, cache=my_cache)
+
+if secrets_singleton.load_if_toml_exists():
+    auth_section = secrets_singleton.get("auth", None)
```

Comment:
With recent changes to secrets, should we assume the auth might update over time? Perhaps for key rotating or something?